### PR TITLE
Add proposal OS API and OS internal Traffic TLS encryption and PKI

### PR DESCRIPTION
**Pull request by fdobrovolny**
https://github.com/SovereignCloudStack/issues/issues/457

### CODE REVIEW

File: Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+title: OpenStack API and OpenStack internal Traffic TLS encryption and PKI Infrastructure
+type: Decision Record
+status: Draft
+track: IaaS
+---
+
+## Introduction
+
+OpenStack is a platform used for the SCS IaaS layer (Infrastructure as a Service).
+Currently, the deployment is mostly done using [Kolla](https://opendev.org/openstack/kolla-ansible).
```
Comment (fkr) [inline Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md:11]:
```suggestion
Currently, the deployment is done via [OSISM](https://github.com/osism/) which makes heavy use of [Kolla](https://opendev.org/openstack/kolla-ansible).
```

File: Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+This ADR document current status of Kolla in regard to end-to-end TLS encryption of selected OpenStack services shown below.
```
Comment (fkr) [inline Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md:12]:
```suggestion
This ADR documents the current status of Kolla in regard to end-to-end TLS encryption of selected OpenStack services shown below.
```

**Comment by fkr** [comment]
@artificial-intelligence can you read this and share your thoughts?

File: Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+2. Horizon
+3. Glance
+4. Cinder
+5. Placement
+6. Nova
+7. Netron
+8. Heat
+13. OpenVSwitch
+
+Only RabbitMQ is not currently supported.
```
Comment (artificial-intelligence) [inline Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md:102]: there is interest in adding an internal CA for rabbitmq and libvirt TLS support, see:

https://etherpad.opendev.org/p/kolla-caracal-ptg#L283

the problematic part currently is, choosing the appropriate CA software, which has many challenges, as can be seen in the linked document where we evaluated various alternatives.

File: Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+[Details](https://docs.openstack.org/kolla-ansible/latest/admin/tls.html#back-end-tls-configuration)
```
Comment (artificial-intelligence) [inline Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md:104]:
```suggestion
[Details](https://docs.openstack.org/kolla-ansible/2023.1/admin/tls.html#back-end-tls-configuration)
```

File: Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+(External)   (public internet) (HAProxy) (OpenStack Networking)   (OpenStack Services)  (OpenStack Networking)  (Data Service)
+   |                                                                  |      |      |                            |         |
+ Client                                                           Keystone  Glance  Placement ...               MariaDB   Redis ...
+```
+
+To provide truly end-to-end traffic encryption within our infrastructure, all communications between services and their respective databases and brokers must be conducted via TLS (Transport Layer Security). This measure not only encrypts the data in transit but also ensures the communication channels are authenticated and integrity-protected.
+
+##### Memcached
+
+Memcached does support TLS; however, this support is only [experimental](https://github.com/memcached/memcached/commit/ee1cfe3bf9384d1a93545fc942e25bed6437d910), and custom build binaries are needed. For our use case, this is a no-go.
```
Comment (artificial-intelligence) [inline Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md:119]: so what do we do with this?

File: Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+##### Redis
+
+Redis supports TLS encryption for client and inter-node communication; however, this has to be enabled during the build and is not feature-enabled by default.
+[Details](https://redis.io/docs/management/security/encryption/)
+
+Kolla currently does not support TLS encryption.
+
+Note: We had issues with `oslo.cache` connecting via TLS to redis, minor upstream change might be needed.
+
```
Comment (artificial-intelligence) [inline Standards/scs-XXXX-v1-encrypted-internal-openstack-traffic-and-pki-infra.md:151]: I miss a paragraph about secure libvirt live migration:

https://docs.openstack.org/nova/latest/admin/secure-live-migration-with-qemu-native-tls.html

**Review by artificial-intelligence — CHANGES_REQUESTED** [review]
I'll add more remarks later.

Libvirt TLS is not discussed at all, so I would really at least like to see a paragraph about it.
I'll also need to check if we are missing other important services.

One thing coming to my mind is e.g. nonvnc transport encryption via vencrypt and binding of the communication channel to the novnc proxyserver, as is discussed here:

https://bugs.launchpad.net/kolla-ansible/+bug/2043709

**Comment by artificial-intelligence** [comment]
another service - which is used in SCS - which is not covered here is cloudkitty, mostly used for billing purposes afaik.

see e.g. this commit to enable internal TLS usage, but the cert is not trusted during bootstrap:

https://review.opendev.org/c/openstack/kolla-ansible/+/866598

in general I would advice to go through each openstack service one-by-one and make a detailed review if the service uses any networking and what is transmitted via which network channels and which of these are already encrypted and if not why not.

So I don't know if this was already done and the above mentioned services where just missed - this would bring up the interesting question why they where missed - or if this has just not happened.

I would strongly encourage to do this, if it hasn't happened yet, because otherwise we have no complete picture and the work done here might be incomplete and ultimately lead to a false sense of security by suggesting that everything is encrypted when it really isn't.

I hope this helps, thank you all for all the work put into this so far!

**Pull request closed without merging**